plan/eliminate_projection.go

```diff
@@ -14,50 +14,215 @@
 package plan
 
 import (
+	"github.com/pingcap/tidb/context"
 	"github.com/pingcap/tidb/expression"
 )
 
-// EliminateProjection eliminates projection operator to avoid the cost of memory copy in the iterator of projection.
-func EliminateProjection(p PhysicalPlan) PhysicalPlan {
-	switch plan := p.(type) {
-	case *Projection:
-		if !projectionCanBeEliminated(plan) {
-			break
-		}
-		child := plan.children[0].(PhysicalPlan)
-		child.SetSchema(plan.Schema())
-		RemovePlan(p)
-		p = EliminateProjection(child)
-	}
-	children := make([]Plan, 0, len(p.Children()))
-	for _, child := range p.Children() {
-		children = append(children, EliminateProjection(child.(PhysicalPlan)))
+// canLogicalProjectionBeEliminated checks if a PROJECTION in a logical plan can be eliminated.
+// Projection p can be eliminated when all the following conditions are satisfied:
+// 1. it's not the root operator
+// 2. canProjectionBeEliminated(p) returns true
+func canLogicalProjectionBeEliminated(p *Projection) bool {
+	if len(p.Parents()) == 0 {
```

Comment:
If the query is `select * from t limit 1`, the parents will also be 0 and can't be removed.